### PR TITLE
Fix FocusScope issues

### PR DESCRIPTION
**Pull request by devongovett**
This fixes a few FocusScope issues found while working on DatePicker.

* In #2139, we added a check if a FocusScope container another nested one to prevent them from clashing. However, this DOM check only checked if one was contained inside another, not that the active element was actually inside it. For example, a focus scope could be a sibling of another focusable element. Focus scopes can also not have the `contain` property set, in which case this also doesn't apply.
* If you didn't have `restoreFocus` enabled, we still handled the Tab key rather than leaving it to the browser.

The fix is to track which the focus scope hierarchy in a tree so we can allow the correct one to handle focus/blur events rather than all of them. This is done in a map, and via context. Moving to a child focus scope from a parent is allowed, but moving from a child to a parent or sibling is not.

### CODE REVIEW

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 1]
```diff
@@ -185,15 +199,15 @@ function useFocusContainment(scopeRef: RefObject<HTMLElement[]>, contain: boolea
   let focusedNode = useRef<HTMLElement>();
 
   let raf = useRef(null);
-  useEffect(() => {
+  useLayoutEffect(() => {
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:202]: Needed so the focus event is registered before elements within the scope are auto focused, or a parent scope will end up handling the event instead and restoring focus back to that scope.

File: packages/@react-spectrum/dialog/test/DialogTrigger.test.js [hunk 2]
```diff
@@ -965,7 +965,7 @@ describe('DialogTrigger', function () {
       expect(outerDialog).toBeVisible();
     }); // wait for animation
     let innerButton = getByTestId('innerButton');
-    userEvent.tab();
+    act(() => innerButton.focus());
```
Comment (devongovett) [inline packages/@react-spectrum/dialog/test/DialogTrigger.test.js:968]: Appears to be a bug in user-event, where if tabbing from an element with `tabindex="-1"` (as in our dialog elements), an incorrect next element is focused.

Comment (solimant) [reply packages/@react-spectrum/dialog/test/DialogTrigger.test.js:968]: Can we add an inline comment here?

**Review by solimant — APPROVED** [review]

**Review by LFDanLu — COMMENTED** [review]
Currently digging as to why, but I noticed that I can no longer open the nested menu in the following story: https://reactspectrum.blob.core.windows.net/reactspectrum/c7fe7f7a340164635fb64ac1e10c808885dccfbe/storybook/index.html?path=/story/dialogtrigger--with-menu-trigger

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 3]
```diff
@@ -264,34 +279,30 @@ function useFocusContainment(scopeRef: RefObject<HTMLElement[]>, contain: boolea
...
-function isFocusScopeInScope(scopes: HTMLElement[]) {
-  return scopes.some(scope => isFocusScopeDirectChild(scope) || isFocusScopeNestedChild(scope));
-}
+  if (parent === ancestor) {
+    return true;
+  }
 
-function isElementInScope(element: Element, scope: HTMLElement[]) {
-  return !isFocusScopeInScope(scope) && scope.some(node => node.contains(element));
+  return isAncestorScope(parent, scope);
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:305]: won't this just return true immediately? i think it's supposed to be this?
```suggestion
  return isAncestorScope(ancestor, parent);
```

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 4]
```diff
@@ -217,17 +231,20 @@ function useFocusContainment(scopeRef: RefObject<HTMLElement[]>, contain: boolea
...
-      let isInAnyScope = isElementInAnyScope(e.target, scopes);
-      if (!isInAnyScope) {
+      // If focusing an element in a child scope of the currently active scope, the child becomes active.
+      // Moving out of the active scope to an ancestor is not allowed.
+      if (!activeScope || isAncestorScope(activeScope, scopeRef)) {
+        activeScope = scopeRef;
+        focusedNode.current = e.target;
+      } else if (scopeRef === activeScope && !isElementInScope(e.target, scopeRef.current)) {
+        // If a focus event occurs outside the active scope (e.g. user tabs from browser location bar),
+        // restore focus to the previously focused node or the first tabbable element in the active scope.
```
Comment (LFDanLu) [inline packages/@react-aria/focus/src/FocusScope.tsx:241]: Did some digging on my previous [comment](https://github.com/adobe/react-spectrum/pull/2223#pullrequestreview-733198966). Looks like the nested menu is actually opening but then immediately closing. What seems to be happening is that `useFocusContainment`'s `scopeRef` doesn't have the Menu's newly added scope and thus judges the menu item focus event to be outside the active scope and not within a child scope of the currently active scope. This means the menu trigger button gets refocused, automatically closing the menu

**Review by LFDanLu — CHANGES_REQUESTED** [review]

**Comment by snowystinger** [comment]
I noticed that the two FocusScope stories behave much better now, it appears though that something isn't quite right in focusscope--keyboard-navigation-inside-portal
Steps are:
Open three modals
Close the second one (so that two unmount)
Now I can click outside and tab into the first input even though a focus scope is still open

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 4 above)
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:241]: Ugh, I had this story working earlier. Will check again

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 5]
```diff
@@ -87,11 +95,17 @@ export function FocusScope(props: FocusScopeProps) {
...
     scopeRef.current = nodes;
-    scopes.add(scopeRef);
+  }, [children, parentScope]);
+
+  useLayoutEffect(() => {
+    scopes.set(scopeRef, parentScope);
     return () => {
+      if (activeScope === scopeRef) {
+        activeScope = parentScope;
+      }
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:105]: I think https://github.com/adobe/react-spectrum/pull/2223#issuecomment-901343206 might be coming from cleanup here
useLayoutEffect will run in the parent, then the child, so the activescope won't get set correctly

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 3 above)
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:305]: Yeah you're right, good catch.

**Review by snowystinger — APPROVED** [review]
lgtm!

**Review by LFDanLu — APPROVED** [review]
LGTM

**Pull request merged**